### DISCUSSION TITLE
Support for (Nested) Projections

### DISCUSSION THREAD

**Discussion by dohorn** [Cosium/spring-data-jpa-entity-graph#69] [Ideas]
First of all, I think this project is a super useful and necessary addition to Spring Data JPA 👍 What I would like to accomplish right now is to use this for nested projections on lazily loaded 1-To-Many Relations.

Doing this without any entity graph usually ends in a LazyInitializationException. However annotating the projection method with an entity graph solves this issue, like so:

```Java
public interface ProductDto {
    String getName();
    List<RatingDto> getRatings();
}

public interface RatingDto {
    int getRating();
}

public interface BaseProductRepository<T extends AbstractProduct> extends Repository<T, String>, EntityGraphQuerydslPredicateExecutor<T> {
    @EntityGraph(attributePaths = {"ratings"})
    Optional<ProductDto> findById(String id);
}
```

As I want to do this in a dynamic way, this project would suit my purposes perfectly, if you decided to add support.

**Comment by reda-alaoui** [comment]
Doesn't it already work with the following code ?

```java
public interface ProductDto {
    String getName();
    List<RatingDto> getRatings();
}

public interface RatingDto {
    int getRating();
}

public interface BaseProductRepository<T extends AbstractProduct> extends Repository<T, String>, EntityGraphQuerydslPredicateExecutor<T> {
    Optional<ProductDto> findById(String id, EntityGraph entityGraph);
}
```

**Comment by dohorn** [comment]
Hmm, so I tried it that way, but first I got a ClassNotFoundException for the class MoreObjects in the class EntityGraphBean. After adding Google Guava to my own POM I get the following Exception:

> com.cosium.spring.data.jpa.entity.graph.repository.exception.InapplicableEntityGraphException: Cannot apply EntityGraph EntityGraphBean{jpaEntityGraph=JpaEntityGraph [name=redacted.model.Product-_-_-_-_-_-, type=LOAD, attributePaths=[ratings]], domainClass=class redacted.model.Product, repositoryMethodReturnType=java.util.Optional<redacted.model.ProductDto>, optional=false} to the the current query
	at com.cosium.spring.data.jpa.entity.graph.repository.support.RepositoryMethodEntityGraphExtractor$JpaEntityGraphMethodInterceptor.doInvoke(RepositoryMethodEntityGraphExtractor.java:145) ~[spring-data-jpa-entity-graph-2.6.0.jar:na]
	at com.cosium.spring.data.jpa.entity.graph.repository.support.RepositoryMethodEntityGraphExtractor$JpaEntityGraphMethodInterceptor.invoke(RepositoryMethodEntityGraphExtractor.java:102) ~[spring-data-jpa-entity-graph-2.6.0.jar:na]

**Comment by reda-alaoui** [comment]
I see.
I created https://github.com/Cosium/spring-data-jpa-entity-graph/issues/70 to implement this feature ;)

**Reply by dohorn** [reply]
Awesome, thanks a lot :)
